File: Python/Calculator/CalculatorServer/Utils/Calculator.py
import calculator_pb2
import calculator_pb2_grpc

class Calculator(object):

    MINIMUM_OPERANDS = 2

    class Operations(object):
        @staticmethod
        def add(first: float, second: float) -> float:
            return first + second;

        @staticmethod
        def substract(first: float, second: float) -> float:
            return first - second;

        @staticmethod
        def multiply(first: float, second: float) -> float:
            return first * second;

        @staticmethod
        def divide(first: float, second: float) -> float:
            return first / second;

    def __init__(self, *args, **kwargs):
        self.__numbers = []

    def add_operand(self, operand: float):
        self.__numbers.append(operand)

    def can_calculate(self):
        return len(self.__numbers) >= self.MINIMUM_OPERANDS

    def calculate(self, operation: calculator_pb2.OperationType):
        return self.__compute(operation)(self.__get_operand(), self.__get_operand()) if self.can_calculate() else float()

    def __compute(self, operation: calculator_pb2.OperationType):
        if operation == calculator_pb2.Addition:
            return self.Operations.add
        if operation == calculator_pb2.Subtraction:
            return self.Operations.substract
        if operation == calculator_pb2.Multiplication:
            return self.Operations.multiply
        if operation == calculator_pb2.Division:
            return self.Operations.divide
        return self.Operations.add

    def __get_operand(self):
        if self.__numbers:
            return self.__numbers.pop()

File: .NET Core 3.0/Calculator/CalculatorServer/Actions/Calculator.cs
using Generated;
using System;
using System.Collections.Generic;

namespace CalculatorServer.Actions
{
  internal sealed class Calculator
  {
    #region Constants
    private const int MinimumOperands = 2;
    private readonly Stack<double> numbers = new Stack<double>();
    #endregion

    #region Properties
    public bool CanCalculate
    {
      get
      {
        return this.numbers.Count >= MinimumOperands;
      }
    }
    #endregion

    #region Public methods
    public void AddOperand(double operand)
    {
      this.numbers.Push(operand);
    }
    public double Calculate(OperationType operationType)
    {
      return CanCalculate ? Compute(operationType)(GetOperand(), GetOperand()) : (default);
    }
    #endregion

    #region Private methods
    private Func<double, double, double> Compute(OperationType operationType)
    {
      switch (operationType)
      {
        case OperationType.Addition:
          return Operations.Add;
        case OperationType.Subtraction:
          return Operations.Substract;
        case OperationType.Multiplication:
          return Operations.Multiply;
        case OperationType.Division:
          return Operations.Divide;
        case OperationType.InvalidOperation:
        default:
          return Operations.Add;
      }
    }
    private double GetOperand()
    {
      try
      {
        return this.numbers.Pop();
      }
      catch
      {
        return default;
      }
    }
    #endregion
  }
}
